### ISSUE TITLE
[FEATURE] output calibration

### ISSUE THREAD

**Issue by sdatkinson** [sdatkinson/NeuralAmpModelerPlugin#529]
## Summary

"Calibrate" the output of a model that has both metadata for both its input and output dBu levels.

- [x] Functionality
- [x] Controls style and layout
- [ ] Disable unavailable options depending upon the model that's loaded

## The math

Suppose...
* Model was trained with a send level of X dBu @ 0 dBFS
* The return calibration is Y dBu @ 0 dBFS
* The plugin host's input is calibrated so that 0 dBFS is Z dBu.

Then, the behavior should be:
1. Increase the gain at the input by (Z-X) dB to calibrate the input.
2. Process
3. Increase the gain of the output by (Y-Z) dB.

(1) achieves input calibration. Because of (3), if the output of the model were a sine wave with 0 dBFS peak, then we know that it is Z dBu.

Example: X=19 Bu (nice and hot send to the amp), Y=10 dBu (e.g. a Focusrite Scarlett with the input gain turned up a tad), and Z=12 dBu (playing through a Focusrite Scarlett with the input trim at minimum). Before the model, the signal is reduced by 7 dB, and after the model the gain is reduced by 2 dB.

This way, if a second NAM is put behind this, then its input calibration can (also) be set to Z dBu and it will be correct to the output of this block.

[Within a plugin that holds two NAMs, the gain-staging between two models with (X1 dBu send, Y1 dBu return) and (X1 send, Y2 return) would be to increase the gain by Y1-X2 dB; this is the same as the two-plugin solution: (Y1-Z) - (Z-X2), Z's cancel. And so on...]

## Controls
In short, change the "normalize" switch to a ternary switch "Output level" with states:
* "Raw" (cf normalize switch off)
* "Normalized" (cf Normalize switch on)
* "Calibrated" (cf this new feature).

_Technically_, setting it to "Calibrated" would be most "sensible" if the _input_ is also calibrated, but it doesn't _need_ the input to be calibrated--that's just like having a signal boost (or cut) hitting the front of the model--just like you'd get if you moved the input level knob on the main page. Besides, having the input calibration switch interact with the output level control would be messy, especially when paging through models if some of them are missing certain metadata. These are things that someone building a product might want to think through, though.

**Comment by sdatkinson** [comment]
Do in on fell swoop with #502.

**Comment by 2-dor** [comment]
Steve, how would this account for scenarios where the reamped signal is not being routed back into the instrument input of the audio interface? 
For instance:
- I shoot the training sequence via my reamp box to the amp (i measure the SEND and RETURN separately like in the readthedocs link you published)
- Amp goes into a loadbox
- from the loadbox I run a Line Out into a Line Input of my audio interface
-

**Comment by sdatkinson** [comment]
@2-dor fair question--people don't usually think about this. It depends whether you're thinking of the load box as part of your signal chain or part of your interface.  I'd argue that neither is necessarily more "right" than the other, though I can imagine how the internet might make up its mind about this 😉 

Either way, the output calibration level is defined as the analog level (in dBu, RMS) that corresponds to 0 dBFS peak in the digital domain after it's converted by your interface.

### Case 1: "It's part of the signal chain"

In this case, measure the level out of the loadbox that clips the interface as you've set the input gain. Nothing different really.

### Case 2: "It's part of the interface"

Ok, so the input to your "interface" is the input to your _loadbox_--meaning the analog level you're trying to quantify is the one that's coming out of the speaker output on your amp. You know, the dangerous one. The level in dBu there would be "off the charts".

## Which one's "right?"

It depends what's after your NAM and how that's been represented. _Usually_, that's an IR, time-based effects, etc--things that most people don't think of as "analog" in this way. In that case, you're probably not going to be using the "calibrated" setting anyways--you'll keep the output mode on "normalized" because that's what makes the models all the same (digital) loudness, which is nice.

_However_, maybe you've made a neural model of your speaker & mic because you're not happy w/ IRs and you're after modeling its nonlinear behavior & such. That breakup happens at an (analog) level--so you (or the person making that cabinet/mic model) needs to basically decide the same thing--is their "input level" either (A) the level coming out of a reamp box (pre- power amp), or is it (B) the power amp output level that's hitting the speaker?) The former is going to be something normal like 15 dBu; the latter is _also_ going to be "Off the charts" just like Case 2 above.

So the punchline is that "Case 1" and "Case A" "go together", and so do "Case 2" and "Case B". But all of this calibration stuff has always been about piees of gear in series "going together", so hopefully that checks out.

---

_All that said_, the output of a tube power amp interacts with the impedance characteristics of the speaker load, so reducing it to level calibration isn't really "accurate" to what's going on anyways. If you chain an amp model to a speaker/mic model, the signal path being modeled (probably) wasn't "amp -> cab -> mic"; it was "amp -> load box -> interface input -> interface output -> reamp box -> power amp -> cab -> mic" 🤷🏻 .

**TLDR: Don't worry about all this when you're modeling your amps. I'm imagining people will find output calibration more immediately useful for _pedal_ models.**

**Comment by 2-dor** [comment]
> @2-dor fair question--people don't usually think about this. It depends whether you're thinking of the load box as part of your signal chain or part of your interface. I'd argue that neither is necessarily more "right" than the other, though I can imagine how the internet might make up its mind about this 😉
> 
> Either way, the output calibration level is defined as the analog level (in dBu, RMS) that corresponds to 0 dBFS peak in the digital domain after it's converted by your interface.
> 
> ### Case 1: "It's part of the signal chain"
> In this case, measure the level out of the loadbox that clips the interface as you've set the input gain. Nothing different really.
> 
> ### Case 2: "It's part of the interface"
> Ok, so the input to your "interface" is the input to your _loadbox_--meaning the analog level you're trying to quantify is the one that's coming out of the speaker output on your amp. You know, the dangerous one. The level in dBu there would be "off the charts".
> 
> ## Which one's "right?"
> It depends what's after your NAM and how that's been represented. _Usually_, that's an IR, time-based effects, etc--things that most people don't think of as "analog" in this way. In that case, you're probably not going to be using the "calibrated" setting anyways--you'll keep the output mode on "normalized" because that's what makes the models all the same (digital) loudness, which is nice.
> 
> _However_, maybe you've made a neural model of your speaker & mic because you're not happy w/ IRs and you're after modeling its nonlinear behavior & such. That breakup happens at an (analog) level--so you (or the person making that cabinet/mic model) needs to basically decide the same thing--is their "input level" either (A) the level coming out of a reamp box (pre- power amp), or is it (B) the power amp output level that's hitting the speaker?) The former is going to be something normal like 15 dBu; the latter is _also_ going to be "Off the charts" just like Case 2 above.
> 
> So the punchline is that "Case 1" and "Case A" "go together", and so do "Case 2" and "Case B". But all of this calibration stuff has always been about piees of gear in series "going together", so hopefully that checks out.
> 
> _All that said_, the output of a tube power amp interacts with the impedance characteristics of the speaker load, so reducing it to level calibration isn't really "accurate" to what's going on anyways. If you chain an amp model to a speaker/mic model, the signal path being modeled (probably) wasn't "amp -> cab -> mic"; it was "amp -> load box -> interface input -> interface output -> reamp box -> power amp -> cab -> mic" 🤷🏻 .
> 
> **TLDR: Don't worry about all this when you're modeling your amps. I'm imagining people will find output calibration more immediately useful for _pedal_ models.**

Definitely - it's a can of worms to be honest & I agree - very useful for pedals mostly. 

I think I got tripped up in reading the notes & wondering how do the measurements carried out for the INPUT calibration tied into this. I think that's a big part for most gear to get right (the INPUT piece) - pedals will benefit from this for sure.
